libopenage/screenshot.cpp

```diff
@@ -46,44 +50,83 @@ std::string ScreenshotManager::gen_next_filename() {
 
 
 void ScreenshotManager::save_screenshot() {
-	std::string filename = this->gen_next_filename();
+	coord::pixel_t width = this->window_size.x,
+	               height = this->window_size.y;
 
-	log::log(MSG(info) << "Saving screenshot to " << filename);
+	std::shared_ptr<uint8_t> pxdata(new uint8_t[4*width*height], std::default_delete<uint8_t[]>());
+	glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pxdata.get());
 
-	int32_t rmask, gmask, bmask, amask;
-	rmask = 0x000000FF;
-	gmask = 0x0000FF00;
-	bmask = 0x00FF0000;
-	amask = 0xFF000000;
+	auto encode_function = [this, pxdata, width, height] () mutable {
+		return this->encode_png(pxdata, width, height);
+	};
+	this->job_manager->enqueue<bool>(encode_function);
+}
 
-	SDL_Surface *screen = SDL_CreateRGBSurface(
-	SDL_SWSURFACE,
-	this->window_size.x,
-	this->window_size.y,
-	32,
-	rmask, gmask, bmask, amask);
 
-	size_t pxcount = screen->w * screen->h;
+bool ScreenshotManager::encode_png(std::shared_ptr<uint8_t> pxdata,
+	                               coord::pixel_t width, coord::pixel_t height) {
+	std::string filename = this->gen_next_filename();
+	FILE *fout = fopen(filename.c_str(), "wb");
```

Comment:
why opening the fd up here, and not after the initialization of libpng - it would make many `fclose(fout)` obsolete.